addons/project/controllers/project_sharing_chatter.py

```diff
@@ -0,0 +1,64 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from werkzeug.exceptions import Forbidden
+
+from odoo.http import request, route
+from odoo.tools import consteq
+
+from odoo.addons.portal.controllers.mail import PortalChatter
+from .portal import ProjectCustomerPortal
+
+
+class ProjectSharingChatter(PortalChatter):
+    def _check_project_sharing_access(self, project_id, res_model, res_id, token):
+        """ Check if the chatter in project sharing can be accessed
+
+            If the portal user is in the project sharing, then we do not have the access token of the task
+            but we can have the one of the project (if the user accessed to the project sharing views via the shared link).
+            So, we need to check if the chatter is for a task and if the res_id is a task
+            in the project shared. Then, if we had the project token and this one is the one in the project
+            then we return the token of the task to continue the portal chatter process.
+            If we do not have any token, then we need to check if the portal user is a follower of the project shared.
+            If it is the case, then we give the access token of the task.
+        """
+        project_sudo = ProjectCustomerPortal._document_check_access(self, 'project.project', project_id, token)
+        can_access = project_sudo and res_model == 'project.task' and res_id in project_sudo.tasks.ids and project_sudo._check_project_sharing_access('comment')
```

Comment:
performance-wise, comparing the `task.project_id` with `project_id` will be much more efficient than `res_id  in project_sudo.tasks.ids`. It may look like more code, but it only requires retrieving a single task row in the database, while the current code requires retrieving all tasks of the project (which may have several thousands of them).

also you want to short-circuit those checks as much as possible, i.e. put the conditions in order from cheapest to most expensive